### DISCUSSION TITLE
How do we get "LastVerified" for a Replica

### DISCUSSION THREAD

**Discussion by hannahhoward** [filecoin-project/motion#146] [Q&A]
Just want to flag that determining the last time a deal was verified from the chain is not super easy. We had probably start thinking about the right Lotus API calls needed to get there.

**Comment by masih** [comment]
A bigger question here is what "verified" means. It can be:
* Deal is published, not expired, not slashed and active on Lotus API the call for which is `StateMarketStorageDeal`, and
* Attempt to retrieve a random range of data from the SP.

I think that's about as best as we can do?

Once that's done, then  `LastVerified` would become an internal timestamp.


Also the FIP below will impact this to some extent and verification should ideally not be strongly tied to "deals":

* https://github.com/filecoin-project/FIPs/discussions/730

**Comment by celine3650** [comment]
I was just catching up notes from the product discussion last week 9/12. 

Based on the discussion, SP wants to put a green mark on the data verifiability as it's the core value prop of filecoin. That means to be able to read date / time of most recent successful verified deals for each blob is critical. 
    1. For each SP and given blob, when was the last time it was verified

**And for each blob, what was the last successful time + most recent time is how we define "verified"** 
@masih @hannahhoward @elijaharita  Do you agree if we use this as an acceptance criteria to determine the "Lastverified"? 

Given the implementation difficulty, a proposed path is to separate diagnostics API that grants Filecoin information for information purposes. Any thoughts on that?

**Comment by masih** [comment]
@celine3650 
* separate diagnostics API sounds reasonable and is already captured as an issue here: https://github.com/filecoin-project/motion/issues/43 
* In the meantime, LastVerified data represents whatever Singularity defines as "Verified", which is the timestamp of a periodic check performed by Singularity that confirms:
  * deal is published
  * deal is active on chain.

**Comment by masih** [comment]
For Beta, scope reduced to #132. More sophisticated checking will be performed in Singularity eventually and Motion will remain a client of singularity in terms of reporting the last verified date. Therefore, removing the beta label from this issue.
